Add coverage testing.

LukeShu opened the pull request:
This patchset adds coverage testing.  After running `make check`, the master coverfile is `test/cov.cov`, and an HTML report is generated from it at `test/cov.html`.

This "vendors" in `util/` two new utilities: `write-ifchanged`, which I've been copying from project to project for several years, and [`gocovcat`](https://git.lukeshu.com/go/cmd/gocovcat/). Though I've since created a separate repository for gocovcat, it actually originated in holo and I git-filter-branch'ed it out.

The most invasive part of this change is that it switches over to use the normal Go workspace structure.  This means that output binaries go in `bin/` instead of `build/`.  Because `build/` no longer exists, I've moved the output location for manpages to `man/`.

Commit:
Save version information to a file, instead passing it as a flag.

In general; I prefer this approach, as it means that things are correctly
rebuilt if the version changes.

However, the push for it being done now is that I don't believe that the
linkerflags-based approach can be made to work with the `go test`-produced
binaries that an upcoming commit will add.

Now, because forcing the file to be generated would break `go get`, the
linkerflags-based approach still works; it just isn't used by the Makefile.
A `go get` user will have an empty version string.

Commit:
Add the ability to create coverage-testing tooled binaries.

At a high-level, how this is done is by defining a TestXXX() function that
calls `main()`; and the "testing" subsystem takes care of everything else.

There are a couple of gotchas with this.

The big gotcha is that we can't call os.Exit(), or the process will exit
before the coverage results have been written.  Unfortunately, we need to
be able to exit with non-zero codes.  So, instead of having our Test
function call `main()`, we add a layer of indirection:

	+func main() {
	+	os.Exit(Main())
	+}
	+
	-func main() {
	+func Main() int {

and have the Test function call Main() instead of main().  However, now we have
to be disciplined and make sure that we *return from Main()* rather than call
os.Exit().  Although, this change isn't actually present in this commit because
I'd previously backported it to fix the leaking cachedir issue... but before a
rebase it was part of this commit.

Another issue is that the "testing" package requires use of the "flag"
package, rather than other argument parsing.  But holo doesn't use "flag".
So, we bypass "flag"'s normal behavior by calling
`flag.CommandLine.Parse(args)` ourselves with the flags we want to pass to
the testing subsystem, which we get from the environment variable
HOLO_TEST_FLAGS.

The final issue is that before exiting, the "testing" subsystem prints a
message informing the user that all tests ran successfully.  This breaks
our binaries, as we care about what happens on stdout.  So, we simply close
os.Stdout after calling Main(), but before returning control to the testing
subsystem.

Oh, and there's one more thing not related to the testing itself: You have
to give `go test` a list of packages that you want it to keep track of
coverage for.  Normally we could used `github.com/holocm/holo/...` to track
everything, but `go` won't follow symlinks when expanding `...` wildcards;
so we use `go list` to enumerate the packages that are actually getting
compiled.

LukeShu commented:
BTW, Adding a coveralls.io badge to the README would be easy, but would require you to create an account and add the repo there.

majewsky commented:
Wow. I will have to take my time to dig through this. Do you have a reference link for the "normal Go workspace structure" for me? I've not seen this convention before.

LukeShu commented:
Perhaps you just haven't heard the term "workspace" before (it seems to be less common); it refers to a directory pointed to by `$GOPATH`.

Workspace reference: https://golang.org/doc/code.html#Workspaces
`go` command reference saying that relative import paths are not allowed in workspaces: https://golang.org/cmd/go/#hdr-Relative_import_paths

majewsky commented on Makefile:
```diff
@@ -1,56 +1,72 @@
-build/man/%: doc/%.pod
+man/%: doc/%.pod | man
 	pod2man --name="$(shell echo $* | cut -d. -f1)" --section=$(shell echo $* | cut -d. -f2) \
 		--center="Configuration Management" --release="Holo $(VERSION)" \
```
The `$(VERSION)` interpolation is broken after you changed the usage of `util/find_version.sh`.

majewsky commented on src/holo-files/main_test.go:
```diff
@@ -0,0 +1 @@
+../main.go.test
```
What is the purpose of this test (and of `src/holo/main_test.go`)?

LukeShu replied:
The binaries produced by `go test` don't actually call `main()`; instead, you need to define a test that calls your main function.

majewsky commented on src/main.go.test:
```diff
@@ -0,0 +1,54 @@
+
+func TestSystem(t *testing.T) {
+	exit = Main()
+	_ = os.Stdout.Close()
```
You had some additional comments about `Main()` and `os.Stdout.Close` in the message for this commit. Can you add these into the code instead?

majewsky reviewed (commented):
Please find below some nitpicks. I will add another comment about the big picture, which doesn't fit in the small textbox that Github offers me right now. :)

majewsky commented:
Big picture issue no. 1: I'm not satisfied with the workspace structure as it is established in this branch. There are a lot of properties that I like, esp. "no relative imports". That was just the younger me not knowing how to do things properly.

I've seen a lot of other people's Go repos in the previous months, and the general pattern for how they organize their code [1] uses directories `<repo-root>/cmd/<binary>` for each binary, and groups the shared code into `<repo-root>/pkg`. Imports are always done as `import "github.com/<org>/<repo>/pkg/<module>"`.

Compared to that established pattern, a line like `import "holo-files/common"` looks really surprising. Out of context, I would wonder why holo-files is part of the stdlib now. Module names that don't comply with the layout of the user's regular GOPATH also shut the door on ever trying to import these modules somewhere else.

I will try to do a competing PR with the repo layout that is floating in my head right now, so we can discuss the merits of both versions against each other.

[1] As an example, see https://github.com/kubernetes/helm.

Big picture issue no. 2: There's few documentation on all the clever contraptions that you designed to make the coverage testing work. I'm willing to merge the PR without docs, though. It will be a nice exercise to write the documentation myself and have you review if I really understood everything. :)

LukeShu commented:
I've definitely got writing a blog post about the coverage testing on my TODO list :)

My implementation was based heavily on [this article from Elastic](https://www.elastic.co/blog/code-coverage-for-your-golang-system-tests), but [this](https://blog.cloudflare.com/go-coverage-with-external-tests/) article would also have been very helpful if it had existed at the time.

Commit:
Use the coverage-testing tooled binaries for running the test suite.

This required adjusting tests 11-fail-scan and 23-alpm-hook, in addition
to the shared holorc (which was enough to adjust all of the other tests).

The final coverage report is placed at `test/cov.cov`, which is built from
separate coverage reports for each invocation at `test/cov/*.cov`.

An HTML report is generated from this at `test/cov.html`.

LukeShu commented:
Oh, if you don't like having the "vendored" `gocovcat` in `util/`, we could add an external dependency on it, or on github.com/wadey/gocovmerge (which as I understand it does the same thing).  Perhaps via a git submodule, if you don't think those are too awful.

majewsky commented:
I'm fine with the vendored gocovcat, following the precedent of `util/find_version.sh` being copied into most other repos in this org.

Re git submodules, I've had quite the negative experience them, esp. because they are not considered by Github when it creates release tarballs. See https://github.com/holocm/holo-attic/blob/master/util/make-release-tarball.sh for the hairs that I pulled out over this.

majewsky commented:
Now that #28 is merged, please rebase.

LukeShu commented:
In rebasing this, I learned that the `go` command doesn't follow symlinks when evaluating `...` wildcards (see `go help packages`, though it doesn't mention links); which I used to list the packages to coverage test for.

So it will take me a bit to work around `$GOPATH/src/github.com/holocm/holo` being a symlink.

LukeShu committed:
Makefile: Use a variable for "github.com/holocm/holo", simplify "clean".

This is to keep the noise in the next commit down.

LukeShu commented:
The command to build the coverage-testing binaries is now pretty gross.  It uses `$(go list)` with some arguments to enumerate the packages that it is using.

LukeShu commented:
Ping?

majewsky commented:
Oh yeah, sorry for dropping the ball.

majewsky commented:
I've also set up coveralls.io and added the badge to the README.
